### ISSUE TITLE
brainstorming - 'out of context' training examples

### ISSUE THREAD

**Issue by dobkeratops** [bbernhard/imagemonkey-core#100]
not a suggestion for the site, just a NN training idea based on 

I was just thinking about how you could train a neural net to get a general sense for scenes;
I notice in LabelMe there's some images named 'out of context' with what appear to be photoshopped nonsensical situations (e.g. people standing in impossible places, etc etc)

Given the 'alpha-channel'/'outline' data (i.e. your smart-annotations or any other traced boundaries), I wonder if these 'out-of-context' images could be generated randomly, giving you an endless training set , and you train a net to determine if they are 'out of context' or not - although you would need the *originals* to ensure it's not just recognising the set of pastable objects, maybe it would take something more elaborate like adjusting it's brightness to match the surroundings too since many 'cut-pastes' would be obvious from lighting. what I'd really be hoping for is the earlier layers of such a net would develop a sense of scale and surfaces (which surfaces support which type of objects)

alpha channel outlines I'd been doing manually (you could generate those from any polygonal annotations...)
<img width="493" alt="screen shot 2018-04-15 at 07 04 06" src="https://user-images.githubusercontent.com/1120754/38775470-3f4ee9ba-407b-11e8-8173-eb1d322bf0fe.png">

.. pasted over real scenes..

![out_of_context](https://user-images.githubusercontent.com/1120754/38775473-5bb46b0c-407b-11e8-81d7-d36087843169.jpeg)

![out_of_context2](https://user-images.githubusercontent.com/1120754/38775477-638670be-407b-11e8-8d51-364ac20ab13e.jpeg)

perhaps it would be possible to do something 'GANN'-like ,train 2 nets adversarially to figure out viable locations given real images and a set of 'sprites' extracted from them..

**Comment by bbernhard** [comment]
Hahaha...that's a nice idea - I really like that. In particular, the idea to auto-generate those images is pretty cool! I wonder if that could be a candidate for gamification (e.q guess if the image is fake or real) or if it takes too long until that gets challenging.

**Comment by dobkeratops** [comment]
I wonder if there is also utility in asking it to recognise the pasted car as a car (whether or not it's in context) - as that might help the net recognise the car more independently of the background (I think with the less clear examples however, the background context *helps it*

If given scenes with further masks painted (e.g. some indication of layering) it might be possible to help it recognise how things are occluded , by pasting *behind* things - train a net to guess where the rest of the car is

furthermore given 'pavement' and 'road' channels, you could accelerate the process by using those as guide to place the pastes (people on pavements, cars on the road, then the contextual training can get straight onto learning about scale and direction)

**Comment by bbernhard** [comment]
> furthermore given 'pavement' and 'road' channels, you could accelerate the process by using those as guide to place the pastes (people on pavements, cars on the road, then the contextual training can get straight onto learning about scale and direction)

good idea! I guess that could be something worth to invest a bit of time to. Not sure though if it's something that should be part of the imagemonkey-core package, but I guess it could fit nicely into the imagemonkey-playground. :)
